### PR TITLE
rac-3802 bring larval logging and stream monitors to public space

### PR DESCRIPTION
**Pull request by stuart-stanley**
Change to common/fit_common.py
* bring in new "flogging" (fit-logging) module from stream-monitor plugin space
* added a merged file to flogging (all_all).
* adjust run_nose() method to:
** fix: not use lucky-scoping to get access to noseopts variable
** converted invocation to use built environment and args set (rather than monster subprocess string)
*** env passed in via official api
*** noseargs passed in the same
*** converted from dangerous string to list-of-args
* added run_from_module() method (see changes to tests/rackhd11/test_rackhd11_api_catalogs.py)

Everything under stream-monitor is basically the plugin from PR https://github.com/stuart-stanley/nose-observer-plugin/pull/1.
It contains four sub-dirs and some support files. This is implemented as a plugin for nose!
* support files:
** readme, mkenv.sh, and requirements.
* sm_plugin:
** the real plugin base. basically just enough smarts right now to orchestrate telling future stream-sources "test starting!" "test ending!" and so on.
* stream-sources: where all the actual stream "sources" will live. At the moment, this contains a (test harness) logging stream monitor:
** smart enough to detect our local-loggers
** able to inject additional log message to bracket test start/stops
** ...which also means those contact points will be available to use a common scanner (did I see XXX? Yeah? GO BOOM!) and so on.
* flogging is the code for:
** rac-263 (basic log handles for use in test areas)
** rac-264 (granular logging levels)
** this code is meant to be used from fit_tests via common/fit_common.py:get_loggers()

tests/rackhd11/test_rackhd11_api_catalogs.py shows an example conversion:
* bring in the multi-loggers holder
* convert a couple of prints to use them:
** use of "non-expert I'm just writing a test" simplified form of logger
** demo the use of debug_0, _1, and _2 (logger levels) vs just "debug"
* convert __name__ module invoker to use new run_from_module to nose this specific file

Stragglers:
* touched up requirements file
* Update Vagrant to use correct mkenv steps
* Update README to use correct mkenv steps.

Tested with:
python run.py --show-plan
python run.py --group="pollers_api2.tests"
python run_tests.py -stack 1 -test util/display_rackhd_nodes.py
python run_tests.py -stack 1 -test deploy/rackhd_stack_init.py -list
python run_tests.py -stack 1 -test tests -group smoke
python run_tests.py -stack 1 -test tests/api-cit/v2_0/nodes_tests_fit.py -list
python run_tests.py -stack 1 -test tests/api-cit/v2_0/nodes_tests_fit.py:NodesTests.test_07_node_catalogs
python run_tests.py -stack 1 -test tests/api-cit/v2_0/nodes_tests_fit.py
python run_tests.py -stack 1 -test tests -group nodes_api2_tests

@RackHD/corecommitters @johren @jimturnquist @stuart-stanley @gpaulos @hohene

### CODE REVIEW

**Comment by stuart-stanley** [comment]
test this please

**Comment by stuart-stanley** [comment]
So the build is failing because the jenkins virtual env step appears to be 1) hand-crafted and not using ./mkenv.sh and 2) is setup to not create the virtualenv within the workspace, and 3) because of that, it breaks relative package builds (for example, a contained nose plugin....)

@gpaulos, I think you have the rights to at least look at the job. Ideally, it should uses the in-tree commands like ./mkenv.sh...

Everyone else can still review until this is resolved, but don't merge!!!

**Comment by stuart-stanley** [comment]
20:35:26 New python executable in /tmp/.venv/bin/python
20:35:26 Installing setuptools, pip...done.
20:35:27 stream-monitor should either be a path to a local project or a VCS url beginning with 

^^^^ this is why it breaks "/tmp/.venv". Huh. I hope this has exclusive run-mode on the build-nodes, or jobs will step on each others virtenvs in any case!
btw, the line from requirements.txt that needs to be supported is:
-e stream-monitor

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

File: test/stream-monitor/flogging/infra_logging.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        matches = {}
+        for file_name in dirs_in_log:
+            m = matcher.match(file_name)
+            if m is not None:
+                key = m.group("run_date")
+                matches[key] = file_name
+
+        keys = sorted(matches.keys(), reverse=True)
+        trim_these = keys[3:]
+        for trim_inx in trim_these:
```
Comment (hohene) [inline test/stream-monitor/flogging/infra_logging.py]: piddly request, can we keep say 10 old log directories?  running my test programs and already want the last set of 8. :-)

Comment (stuart-stanley) [inline test/stream-monitor/flogging/infra_logging.py]: Sure, but it's the only change you get to request :). (I was going to make it an env variable, but decided there will be other things we want to configure on this and it made sense to not keep adding stuff just now...)

Comment (hohene) [inline test/stream-monitor/flogging/infra_logging.py]: Ha! sounds good!   I have run through many of our sample scripts and all look good so far.  I'll have a few more updates to formatting later, but, want to get more run time with logging before suggesting changes.

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by larry-dean** [comment]
+1

**Comment by jimturnquist** [comment]
+1

**Comment by hohene** [comment]
+1 with a note to add additional log capturing from the nosetests user console output into the console-caputre.log and other logs.   Jenkins captures this data for us automatically.   The request is to also capture this when the user runs the test scripts manually. :-)

**Comment by gpaulos** [comment]
Got a good smoke test run.
Note that building the virtual environment took a long time, like 2 minutes. This will add to test time.
Could not test the deployment stuff because RackHD repo is currently broken.
I think its good! Will test deployment later. If there are any problems we'll address it then.
+1

**Comment by stuart-stanley** [comment]
@gpaulos there is a change into build-config (already merged, but for some reason the Jenkins job is set up to use a private fork). When that gets ironed out, the mkenv will occur during the same interval the test script normally sits a sleep-loops waiting for the RackHD api to come up in vagrant. That step still takes longer than the mkenv.

**Comment by johren** [comment]
test this please

**Pull request merged**